circle/src/domain.rs

```diff
@@ -0,0 +1,359 @@
+        coset0.interleave(coset1).take(1 << self.log_n)
+    }
+
+    #[instrument(skip_all, fields(log_n = %self.log_n))]
+    pub(crate) fn lagrange_basis<EF: ExtensionField<F>>(&self, point: Complex<EF>) -> Vec<EF> {
+        let domain = self.points().collect_vec();
+
+        // the denominator so that the lagrange basis is normalized to 1
+        // this depends only on domain, so should be precomputed
+        let lagrange_normalizer: Vec<F> = domain
+            .iter()
+            .map(|p| s_p_at_p(p.real(), p.imag(), self.log_n))
+            .collect();
+
+        let basis = domain
+            .into_iter()
+            .zip(&lagrange_normalizer)
+            .map(|(p, &ln)| {
+                // ext * base
+                v_0(p.conjugate().rotate(point)) * ln
+            })
+            .collect_vec();
+
+        batch_multiplicative_inverse(&basis)
+    }
+}
+
+impl<F: ComplexExtendable> PolynomialSpace for CircleDomain<F> {
+    type Val = F;
+
+    fn size(&self) -> usize {
+        1 << self.log_n
+    }
+
+    fn first_point(&self) -> Self::Val {
+        point_to_univariate(self.shift).unwrap()
+    }
+
+    fn next_point<Ext: ExtensionField<Self::Val>>(&self, x: Ext) -> Option<Ext> {
+        // Only in standard position do we have an algebraic expression to access the next point.
+        if self.is_standard() {
+            let gen = F::circle_two_adic_generator(self.log_n);
+            Some(point_to_univariate(gen.rotate(univariate_to_point(x))).unwrap())
+        } else {
+            None
+        }
+    }
+
+    fn create_disjoint_domain(&self, min_size: usize) -> Self {
+        let mut log_n = log2_ceil_usize(min_size);
+        // Right now we simply guarantee the domain is disjoint by returning a
+        // larger standard position coset, which is fine because we always ask for a larger
+        // domain. If we wanted good performance for a disjoint domain of the same size,
+        // we could change the shift. Also we could support nonstandard twin cosets.
+        assert!(
+            self.is_standard(),
+            "create_disjoint_domain not currently supported for nonstandard twin cosets"
+        );
+        while log_n == self.log_n {
+            log_n += 1;
+        }
+        Self::standard(log_n)
+    }
+
+    fn zp_at_point<Ext: ExtensionField<Self::Val>>(&self, point: Ext) -> Ext {
+        v_n(univariate_to_point(point).real(), self.log_n) - v_n(self.shift.real(), self.log_n)
+    }
+
+    fn selectors_at_point<Ext: ExtensionField<Self::Val>>(
+        &self,
+        point: Ext,
+    ) -> LagrangeSelectors<Ext> {
+        let zeroifier = self.zp_at_point(point);
+        let p = univariate_to_point(point);
+        LagrangeSelectors {
+            is_first_row: zeroifier / v_0(self.shift.conjugate().rotate(p)),
+            is_last_row: zeroifier / v_0(self.shift.rotate(p)),
+            // is_transition: v_0(self.shift.rotate(p)),
```

Comment:
nit: to be removed ?

Reply:
I suppose not yet.  
@rjeli as mentioned on discord, we should take `1 - Lagrange(last_row, point)` as transition selector, in order to stay within the FFT space. 
Btw. just for clarification: there is no transition selector in the paper. :)